src/NuGet.Core/NuGet.PackageManagement/PackageDownloader.cs

```diff
@@ -49,62 +49,78 @@ public static class PackageDownloader
             var failedTasks = new List<Task<DownloadResourceResult>>();
             var tasksLookup = new Dictionary<Task<DownloadResourceResult>, SourceRepository>();
 
-            using (var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(token))
-            {
-                // Create a group of local sources that will go first, then everything else.
-                var groups = new Queue<List<SourceRepository>>();
-                var localGroup = new List<SourceRepository>();
-                var otherGroup = new List<SourceRepository>();
+            var linkedTokenSource = CancellationTokenSource.CreateLinkedTokenSource(token);
+            // Create a group of local sources that will go first, then everything else.
+            var groups = new Queue<List<SourceRepository>>();
+            var localGroup = new List<SourceRepository>();
+            var otherGroup = new List<SourceRepository>();
 
-                foreach (var source in sources)
+            foreach (var source in sources)
+            {
+                if (source.PackageSource.IsLocal)
                 {
-                    if (source.PackageSource.IsLocal)
-                    {
-                        localGroup.Add(source);
-                    }
-                    else
-                    {
-                        otherGroup.Add(source);
-                    }
+                    localGroup.Add(source);
+                }
+                else
+                {
+                    otherGroup.Add(source);
                 }
+            }
+
+            groups.Enqueue(localGroup);
+            groups.Enqueue(otherGroup);
+
+            while (groups.Count > 0)
+            {
+                token.ThrowIfCancellationRequested();
```

Comment:
This line will prevent the dispose from ever happening if cancellation happened earlier, and still having a memory leak around